protowhat/Feedback.py

```diff
@@ -1,54 +1,93 @@
-        if state is not None:
-            self.highlight = getattr(state, "highlight", None)
-            self.highlighting_disabled = state.highlighting_disabled
-            if hasattr(state, "path"):
-                self.path = state.path
-
-    def _highlight_data(self) -> Dict[str, int]:
-        highlight = {}
-        if self.path:
-            highlight["path"] = self.path
-
-        if hasattr(self.highlight, "get_position"):
-            position = self.highlight.get_position()
-            if position:
-                highlight.update(position)
-        elif getattr(self.highlight, "first_token", None) and getattr(
-            self.highlight, "last_token", None
-        ):
-            # used by pythonwhat
-            # a plugin+register system would be better
-            # if many different AST interfaces exist
-            highlight.update(
-                {
-                    "line_start": self.highlight.first_token.start[0],
-                    "column_start": self.highlight.first_token.start[1],
-                    "line_end": self.highlight.last_token.end[0],
-                    "column_end": self.highlight.last_token.end[1],
-                }
-            )
-
-        return highlight
-
-    def get_highlight_data(self) -> Dict[str, int]:
-        result = None
-        if self.highlight is not None and not self.highlighting_disabled:
-            result = self._highlight_data()
-
-        return result or {}
+class FeedbackComponent:
+    def __init__(self, feedback: str, kwargs=None, append=True):
+        self.feedback = feedback
+        self.kwargs = kwargs or {}
+        self.append = append
 
     def __repr__(self):
         return "<{} {}>".format(self.__class__.__name__, repr(vars(self)))
 
 
-class InstructorError(Exception):
-    pass
+class Feedback:
+    # This is the offset for ANTLR
+    ast_highlight_offset = {"column_start": 1, "column_end": 1}
+
+    def __init__(
+        self,
+        conclusion: FeedbackComponent,
+        context_components: List[FeedbackComponent] = None,
+        highlight=None,
+        path=None,
+        highlighting_disabled=False,
+        highlight_offset=None,
+        full_code_position=None,
+    ):
+        self.conclusion = conclusion
+        self.context_components = context_components or []
+        self.highlight = highlight
+        self.path = path
+        self.highlighting_disabled = highlighting_disabled
+        self.highlight_offset = highlight_offset or {}
+        self.full_code_position = full_code_position
+
+    @classmethod
+    def get_highlight_position(cls, highlight) -> Dict[str, int]:
+        if hasattr(highlight, "get_position"):
+            return highlight.get_position()
+
+    def get_highlight(self) -> Dict[str, int]:
+        highlight = Counter()
+
+        if not self.highlighting_disabled:
+            position = self.get_highlight_position(self.highlight)
+
+            # TODO: disable highlighting everything
```

Comment:
Are we sure we want to disable highlighting everything, the feedback message will still be "Check the highlighted code".